### ISSUE TITLE
Nested wrapper subclasses with torch.compile is broken

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#124878]
This came in in the FSDP2 workstream, which needs a DTensor that holds some sort of float8 tensor (cc @Chillee @ezyang @zou3519 @albanD @samdow @msaroufim @anijain2305 @chauhang @awgu / @drisspg ).

We should probably add general testing that nested wrapper subclasses work properly in compile, as part of fixing that issue. The first thing we should probably do is audit all of the places that we call `__tensor_flatten__` in dynamo/AOTAutograd, and make sure that we are recursively calling `__tensor_flatten__` when the inner tensors are also wrapper subclasses.

In particular, I tried a simple repro using TwoTensor's, and noticed that the gradients we produce with compile are wrong (this runs properly when you remove the compile decorator):
```
import torch
from torch.testing._internal.two_tensor import TwoTensor

@torch.compile(backend='aot_eager')
def f(x):
    return x.sin().cos()

a = torch.ones(4, requires_grad=True)
a2 = a.clone().detach().requires_grad_()
aa = TwoTensor(a, a2)
aa2 = aa.clone().detach().requires_grad_()
aaaa = TwoTensor(aa, aa2)
out = f(aaaa)
print(type(out))
print(type(out.a))
print(type(out.b))
print(type(out.a.a))
print(type(out.a.b))
print(type(out.b.a))
print(type(out.b.b))
out.sum().backward()
# aaaa.grad should be a TwoTensor(TwoTensor, TwoTensor)
# but instead it is a TwoTensor(tensor, tensor)
print(type(aaaa.grad))
print(type(aaaa.grad.a))
print(type(aaaa.grad.b))
print(type(aaaa.grad.a.a))
print(type(aaaa.grad.a.b))
print(type(aaaa.grad.b.a))
```
